Sources/_SubprocessCShims/process_shims.c

```diff
@@ -266,82 +427,177 @@ int _subprocess_fork_exec(
     }
 #endif
 
-    pid_t child_pid = fork();
-    if (child_pid != 0) {
-        *pid = child_pid;
-        return child_pid < 0 ? errno : 0;
+    // Setup pipe to catch exec failures from child
+    int pipefd[2];
+    if (pipe(pipefd) != 0) {
+        return errno;
+    }
+    // Set FD_CLOEXEC so the pipe is automatically closed when exec succeeds
+    short flags = fcntl(pipefd[0], F_GETFD);
+    if (flags == -1) {
+        close(pipefd[0]);
+        close(pipefd[1]);
+        return errno;
+    }
+    flags |= FD_CLOEXEC;
+    if (fcntl(pipefd[0], F_SETFD, flags) == -1) {
+        close(pipefd[0]);
+        close(pipefd[1]);
+        return errno;
     }
 
-    if (working_directory != NULL) {
-        if (chdir(working_directory) != 0) {
-            return errno;
-        }
+    flags = fcntl(pipefd[1], F_GETFD);
+    if (flags == -1) {
+        close(pipefd[0]);
+        close(pipefd[1]);
+        return errno;
+    }
+    flags |= FD_CLOEXEC;
+    if (fcntl(pipefd[1], F_SETFD, flags) == -1) {
+        close(pipefd[0]);
+        close(pipefd[1]);
+        return errno;
     }
 
+    // Finally, fork
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated"
+    pid_t childPid = fork();
+#pragma GCC diagnostic pop
+    if (childPid == -1) {
+        close(pipefd[0]);
+        close(pipefd[1]);
+        return errno;
+    }
 
-    if (uid != NULL) {
-        if (setuid(*uid) != 0) {
-            return errno;
+    if (childPid == 0) {
+        // Child process
+        close(pipefd[0]);  // Close unused read end
+
+        // Perform setups
+        if (working_directory != NULL) {
+            if (chdir(working_directory) != 0) {
+                int error =  errno;
+                write(pipefd[1], &error, sizeof(error));
```

Comment:
we're failing to communicate _what_ went wrong here. If this fails with `ENOENT` it's indistinguishable from `execve` failing with `ENOENT`.